core/chains/evm/headtracker/heads.go

```diff
@@ -94,20 +126,31 @@ func (h *heads) AddHeads(historyDepth uint, newHeads ...*evmtypes.Head) {
 		return heads[i].Number > heads[j].Number
 	})
 
-	// cut off the oldest
-	if uint(len(heads)) > historyDepth {
-		heads = heads[:historyDepth]
+	// yeah, we could have used binarySearch here, but the code was much longer and more complex and did not
+	// solve any performance issues
```

Comment:
You can move this check inside the first loop when creating the map and copy only heads that have equal or higher block number than `minBlockToKeep`. This way you can remove this loop.

```suggestion
	if headCopy.Number >= minBlockToKeep {
		headsMap[head.Hash] = &headCopy
	}
```